services/scribe/backfill/fetcher.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/synapsecns/sanguine/services/scribe/config"
+
+	"github.com/jpillora/backoff"
+)
+
+// RangeFilter pre-fetches filter logs into a channel in deterministic order.
+type RangeFilter struct {
+	// iterator is the chunk iterator used for the range.
+	iterator util.ChunkIterator
+	// for logging
+	startBlock *big.Int
+	// for logging
+	endBlock *big.Int
+	// logs is a channel with the filtered ahead logs. This channel is not closed
+	// and the user can rely on the garbage collection behavior of RangeFilter to remove it.
+	logs chan []types.Log
+	// backend is the ethereum backend used to fetch logs.
+	backend ScribeBackend
+	// contractAddress is the contractAddress that logs are fetched for.
+	contractAddress ethCommon.Address
+	// doneChan is a channel that is closed when the RangeFilter has completed.
+	// this is only to be used by external callers
+	doneChan chan bool
+	// chainConfig holds the chain config (config data for the chain)
+	chainConfig *config.ChainConfig
+}
+
+// bufferSize is how many ranges ahead should be fetched.
+const bufferSize = 15
+
+// minBackoff is the minimum backoff period between requests.
+var minBackoff = 1 * time.Second
+
+// maxBackoff is the maximum backoff period between requests.
+var maxBackoff = 10 * time.Second
+
+// NewRangeFilter creates a new filtering interface for a range of blocks. If reverse is not set, block heights are filtered from start->end.
+func NewRangeFilter(address ethCommon.Address, backend ScribeBackend, startBlock, endBlock *big.Int, chainConfig *config.ChainConfig) *RangeFilter {
+	// The ChunkIterator is inclusive of the start and ending block resulting in potentially confusing behavior when
+	// setting the range size in the config. For example, setting a range of 1 would result in two blocks being queried
+	// instead of 1. This is accounted for by subtracting 1.
+	chunkSize := int(chainConfig.GetLogsRange) - 1
+	return &RangeFilter{
+		iterator:        util.NewChunkIterator(startBlock, endBlock, chunkSize, true),
+		startBlock:      startBlock,
+		endBlock:        endBlock,
+		logs:            make(chan []types.Log, bufferSize),
+		backend:         backend,
+		contractAddress: address,
+		doneChan:        make(chan bool),
+		chainConfig:     chainConfig,
+	}
+}
+
+// closeOnDone closes the done channel when the process is finished.
+func (f *RangeFilter) closeOnDone() {
+	f.doneChan <- true
+}
+
+// GetChunkArr gets the appropriate amount of block chunks (getLogs ranges).
+func (f *RangeFilter) GetChunkArr() (chunkArr []*util.Chunk) {
+	for i := uint64(0); i < f.chainConfig.GetLogsBatchAmount; i++ {
+		chunk := f.iterator.NextChunk()
+		if chunk == nil {
+			return chunkArr
+		}
+		chunkArr = append(chunkArr, chunk)
+
+		// Stop appending chunks if the max height of the current chunk exceeds the concurrency threshold
+		if chunk.EndBlock.Uint64() > f.endBlock.Uint64()-f.chainConfig.ConcurrencyThreshold {
+			return chunkArr
+		}
+	}
+	return chunkArr
+}
```

Comment:
Previously, a chunk size and sub chunk size was specified in the config. This was convoluted and confusing.

Now you specify the amount of get logs requests you want to batch and the range of each. This function creates an array of block ranges for each getlogs request (using an iterator). The batching backend was slightly refactored to take an array of chunks instead of making another iterator.

Reply:
the real fix to the concurrency limits has to come in server.go. Namely, we need to make sure logs we get are less than the lastIndexed for the given contract